doc/design/listener.md

```diff
@@ -0,0 +1,316 @@
+be **event driven** and not a mixture of event and state driven, depending on
+which event is attached, like in the _WaitSet_.
+
+### Requirements
+
+- Whenever an event occurs the corresponding callback should be called **once**
+    as soon as possible.
+- If an event occurs multiple times before the callback was called, the callback
+    should be called **once**.
+- If an event occurs while the callback is being executed the callback should be
+    called again **once**.
+- Thread-safety: attaching an event at any time from any thread.
+- Thread-safety: detaching an event at any time from any thread.
+  - If the callback is currently running `detachEvent` blocks until the callback
+      is finished.
+  - After the `detachEvent` call the event-callback is not called anymore, even
+      when the event was signalled while `detachEvent` was running and the
+      callback was not yet executed.
+  - If the callback is detached from within the event-callback then `detachEvent`
+     is non blocking. The event is detached right after the `detachEvent` call.
+- One can attach at most one callback to a specific event of a specific object.
+  - Usually defined with an enum by the developer. One example is `SubscriberEvent::DATA_RECEIVED`.
+  - Attaching a callback to an event where a callback has been already attached
+      results in an error.
+- One can attach the same event to different objects at the same time.
+- One can attach multiple different events to a single object
+- When the Listener goes out of scope it detaches itself from every attached
+    object via a callback provided by the attached object (like in the WaitSet).
+- When the class which is attached to the Listener goes out of scope it
+    detaches itself from the Listener via a callback provided by the Listener
+    (like in the WaitSet).
+
+### Solution
+
+#### Class diagram
+
+```
+                                   +---------------------------+
+                                   | ConditionVariableData     |
+                                   |   - m_semaphore           |
+                                   |   - m_runtimeName         |
+                                   |   - m_toBeDestroyed       |
+                                   |   - m_activeNotifications |
+                                   +---------------------------+
+                                        | 1               | 1
+                                        |                 |
+                                        | n               | n
++-----------------------------------------------+ +--------------------------------------------------+
+| ConditionListener                             | | ConditionNotifier                                |
+|   ConditionListener(ConditionVariableData & ) | |   ConditionNotifier(ConditionVariableData &,     |
+|                                               | |                 uint64_t notificationIndex)      |
```

Comment:
Is the 1-n cardinality for the `ConditionListener` correct? With this approach events would be distributed across listeners and not every listener would receive all events. While it makes sense so have multiple `ConditionNotifier`, there should be only one `ConditionListener` ... at least semantically.

Reply:
This is wrong. Thanks for the hint!